### PR TITLE
Use synthetic ranges when rewriting range expressions

### PR DESCRIPTION
**Pull request by auduchinok**
Fixes `op_Range` symbol in FCS has non synthetic range that covers the whole `for` expression. In other similar cases the produced ranges are synthetic.

```fsharp
let _ = seq {
    for _ = 0 to 5 do yield! []
}
```

### CODE REVIEW

File: src/Compiler/Checking/CheckComputationExpressions.fs [hunk 1]
```diff
@@ -81,7 +81,8 @@ let (|JoinRelation|_|) cenv env (expr: SynExpr) =
 
     | _ -> None
 
-let elimFastIntegerForLoop (spFor, spTo, id, start, dir, finish, innerExpr, m) = 
+let elimFastIntegerForLoop (spFor, spTo, id, start, dir, finish, innerExpr, m: range) =
+    let m = m.MakeSynthetic()
     let pseudoEnumExpr = 
         if dir then mkSynInfix m start ".." finish
```
Comment (dsyme) [inline src/Compiler/Checking/CheckComputationExpressions.fs:87]: It feels like this use of `m` here and on next line is just wrong.  The range for the `..` expression should be `unionRanges start.Range finish.Range`?

File: src/Compiler/Checking/CheckExpressions.fs [hunk 2]
```diff
@@ -6100,10 +6100,12 @@ and RewriteRangeExpr synExpr =
     match synExpr with
     // a..b..c (parsed as (a..b)..c )
     | SynExpr.IndexRange(Some (SynExpr.IndexRange(Some synExpr1, _, Some synStepExpr, _, _, _)), _, Some synExpr2, _m1, _m2, mWhole) ->
+        let mWhole = mWhole.MakeSynthetic()
```
Comment (dsyme) [inline src/Compiler/Checking/CheckExpressions.fs:6103]: This range probably doesn't need to be made synthetic does it?  I mean, it should be accurate shouldn't it?

File: src/Compiler/Checking/CheckExpressions.fs [hunk 3]
```diff
@@ -6100,10 +6100,12 @@ and RewriteRangeExpr synExpr =
...
         Some (mkSynTrifix mWhole ".. .." synExpr1 synStepExpr synExpr2)
     // a..b
     | SynExpr.IndexRange (Some synExpr1, mOperator, Some synExpr2, _m1, _m2, mWhole) ->
         let otherExpr =
+            let mWhole = mWhole.MakeSynthetic()
```
Comment (dsyme) [inline src/Compiler/Checking/CheckExpressions.fs:6108]: Likewise here

**Review by dsyme — COMMENTED** [review]
Left some comments

File: src/Compiler/Checking/CheckComputationExpressions.fs (hunk 1 above)
Comment (auduchinok) [reply src/Compiler/Checking/CheckComputationExpressions.fs:87]: I think the best option is to cover `..` only, if we have its range in the tree.

But if we make the range synthetic, it'll be ignored by the most of the things anyway, similar to other rewritten expressions.

File: src/Compiler/Checking/CheckExpressions.fs (hunk 2 above)
Comment (auduchinok) [reply src/Compiler/Checking/CheckExpressions.fs:6103]: I found that some/many other cases use synthetic ranges when rewriting expressions, and it's a good way to distinguish the user code from the compiler-generated, which we need in tooling.

Another option here is to produce two distinct non-synthetic ranges that would only cover the `..`. @dsyme What do you think?

File: src/Compiler/Checking/CheckComputationExpressions.fs (hunk 1 above)
Comment (auduchinok) [reply src/Compiler/Checking/CheckComputationExpressions.fs:87]: In this case it's about cases like this:
```fsharp
let a = []
let _ = seq {
    for _ = 1 to 16 do yield! a
}
```
We don't have any `..` here, so I'm not sure what would be a better range to produce?

**Review by T-Gro — DISMISSED** [review]

File: src/Compiler/Checking/CheckComputationExpressions.fs (hunk 1 above)
Comment (dsyme) [reply src/Compiler/Checking/CheckComputationExpressions.fs:87]: The range over `1 to 16` or `1 downto 16` ?

**Comment by dsyme** [comment]
@auduchinok I'll mark as approved (though it's possible me changes to your PR will cause tests to fail...)

File: src/Compiler/Checking/CheckComputationExpressions.fs [hunk 4]
```diff
@@ -81,10 +81,11 @@ let (|JoinRelation|_|) cenv env (expr: SynExpr) =
 
     | _ -> None
 
-let elimFastIntegerForLoop (spFor, spTo, id, start, dir, finish, innerExpr, m) = 
+let elimFastIntegerForLoop (spFor, spTo, id, start: SynExpr, dir, finish: SynExpr, innerExpr, m: range) =
+    let mOp = (unionRanges start.Range finish.Range).MakeSynthetic()
```
[the remark this answers is not in the document]
Comment (auduchinok) [reply src/Compiler/Checking/CheckComputationExpressions.fs:85]: @dsyme Thanks! I'll update the tests if needed.

**Review by T-Gro — APPROVED** [review]

**Pull request merged**